component-catalog/index.html

```diff
@@ -18,56 +18,149 @@
 
       // HIGHLIGHTER
 
-      // start listening to events when the user starts dragging/clickingdown
-      app.ports.highlighterListen.subscribe(function (id) {
+      app.ports.highlighterInit.subscribe(([id, scrollFriendly, readonly]) => {
+        console.log("highlighterInit", id, scrollFriendly, readonly);
+
+        // TODO: do we need to prevent multiple initializations?
+
+        const touchMoveListener = handleTouchEvents("move", id);
         window.requestAnimationFrame(() => {
-          var highlighter = document.getElementById(id);
+          const highlighter = document.getElementById(id);
           if (!id || !highlighter) {
-            // in a real application, report an error at this point.
+            console.log("highlighter not found", id);
             return;
           }
 
-          document.addEventListener("mouseup", onDocumentUp(id));
-          document.addEventListener("touchend", onDocumentUp(id));
+          highlighter.addEventListener("mousedown", (_) => {
+            document.addEventListener("mouseup", onDocumentUp(id, "mouse"), {
```

Comment:
I don't see any `removeEventListener` for the mouseup and mousedown, only the touchstart and touchend.
Also would be nice to have a code comment that "listeners are removed later in handleTouchStartScrollFriendly" or something like that.

Reply:
> I don't see any removeEventListener for the mouseup and mousedown, only the touchstart and touchend.

There's a few different things being confused here, and to clarify I'll explain the number of different things this port does:
1. Detects mouse release outside highlighter area
2. Detects touch release outside highlighter area
3. Detects long-press
4. Handles drag-to-select in a touch device

`mousedown` supports 1. We don't want to remove it. We always want it to register a fire-once `mouseup` that can let us know whether the user has released the mouse button outside the highlighter area. The `mouseup` doesn't need to be removed bc it fires exactly once (see the `once: true` argument).

`touchstart` supports 2 and is also not removed. Same reason as `mousedown`. This `touchend` is also automatically removed bc it fires once, same as `mouseup`.

The events we remove manually are in 3. and 4., for different reasons.

For 4., we remove the event listener to prevent unnecessary port messages from hitting elm in interactions that are not drag-to-select. Highlighter makes TR pages slow, and this helps alleviate some of it.

For 3., we remove event listeners bc they have an "expiry date": once the long-press has registered, we don't need to watch out for movement.

> Also would be nice to have a code comment that "listeners are removed later in handleTouchStartScrollFriendly" or something like that.

The listeners that are removed in `handleTouchStartScrollFriendly` are only the ones created in `handleTouchStartScrollFriendly`. Adding and removing them is right next to each other.

Reply:
I've incorporated some of this explanation into the code. Hopefully it helps future readers of this port!